{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40d3cf98-2f37-46b9-a606-af61ebef2881",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import scipy.io\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import h5py\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata \n",
    "import seaborn as sns\n",
    "from scipy.stats import zscore\n",
    "import matplotlib.pyplot as plt\n",
    "import collections\n",
    "from natsort import natsorted\n",
    "\n",
    "from scipy import stats\n",
    "from scipy import sparse\n",
    "from sklearn.decomposition import PCA\n",
    "from umap import UMAP\n",
    "\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "from scroutines.config_plots import *\n",
    "from scroutines import powerplots # .config_plots import *\n",
    "from scroutines import pnmf\n",
    "from scroutines import basicu\n",
    "from scroutines.gene_modules import GeneModules  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e40623c6-384a-49e4-935d-185e06cfb547",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "import subprocess\n",
    "\n",
    "# run bed to fasta - prepare for motif enrichment analysis\n",
    "        \n",
    "wkdir = \"/u/home/f/f7xiesnm/project-zipursky/v1-bb/v1/results_atac\"\n",
    "f_genome = \"mm10.fna\"\n",
    "\n",
    "# for i in range(5):\n",
    "#     print(i)\n",
    "#     f_bed = f'./L23_regions/L23A_regions_M{i+1}.bed'\n",
    "#     f_fasta = f_bed.replace('.bed', '.fasta') \n",
    "\n",
    "#     # !bed2fasta -both -o test.fasta P10_A_peaks.bed mm10.fna\n",
    "#     result = subprocess.run([\"bed2fasta\", \"-both\", \n",
    "#                                 \"-o\", f_fasta, \n",
    "#                                 f_bed, f_genome,\n",
    "#                                ], cwd=wkdir, stderr=subprocess.PIPE, text=True)\n",
    "#     print(f\"Return code: {result.returncode}\")\n",
    "\n",
    "for i in range(5):\n",
    "    print(i)\n",
    "    f_bed = f'./L23_regions/L23C_regions_M{i+1}.bed'\n",
    "    f_fasta = f_bed.replace('.bed', '.fasta') \n",
    "\n",
    "    # !bed2fasta -both -o test.fasta P10_A_peaks.bed mm10.fna\n",
    "    result = subprocess.run([\"bed2fasta\", \"-both\", \n",
    "                                \"-o\", f_fasta, \n",
    "                                f_bed, f_genome,\n",
    "                               ], cwd=wkdir, stderr=subprocess.PIPE, text=True)\n",
    "    print(f\"Return code: {result.returncode}\")\n",
    "    \n",
    "for i in range(3):\n",
    "    print(i)\n",
    "    f_bed = f'./L23_regions/L23B_regions_M{i+1}.bed'\n",
    "    f_fasta = f_bed.replace('.bed', '.fasta') \n",
    "\n",
    "    # !bed2fasta -both -o test.fasta P10_A_peaks.bed mm10.fna\n",
    "    result = subprocess.run([\"bed2fasta\", \"-both\", \n",
    "                                \"-o\", f_fasta, \n",
    "                                f_bed, f_genome,\n",
    "                               ], cwd=wkdir, stderr=subprocess.PIPE, text=True)\n",
    "    print(f\"Return code: {result.returncode}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f65bdbbf-c455-48c3-8ebd-9719ba9fd8ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # fimo JASPAR2022_CORE_non-redundant_v2.meme test.fasta\n",
    "\n",
    "# # FIMO - scans a set of sequences for *individual matches* to each of the motifs you provide\n",
    "# # fimo --oc fimo_out JASPAR2022_CORE_non-redundant_v2.meme P6_A_peaks.fasta\n",
    "# # this takes long\n",
    "\n",
    "\n",
    "# # AME - analysis of motif enrichment (case vs control)]\n",
    "# # ame --oc ame_out --control P6_C_peaks.fasta P6_A_peaks.fasta JASPAR2022_CORE_non-redundant_v2.meme\n",
    "# ts = [6,8,10,12,14,17,21]\n",
    "# for ti in ts:\n",
    "#     print(ti)\n",
    "#     wkdir = \"/u/home/f/f7xiesnm/project-zipursky/v1-bb/v1/results_atac\"\n",
    "#     f_motifs = 'JASPAR2022_CORE_non-redundant_v2.meme'\n",
    "    \n",
    "#     ame_out = f'ame_P{ti}_C'\n",
    "#     f_test = f'P{ti}_C_peaks.fasta' \n",
    "#     f_cntr = f'P{ti}_A_peaks.fasta' \n",
    "#     result = subprocess.run([\"ame\", \"--oc\", ame_out,\n",
    "#                                  \"--control\", f_cntr, f_test, f_motifs,\n",
    "#                                ], cwd=wkdir, stderr=subprocess.PIPE, text=True)\n",
    "#     print(f\"Return code: {result.returncode}\")\n",
    "\n",
    "#     ame_out = f'ame_P{ti}_A'\n",
    "#     f_test = f'P{ti}_A_peaks.fasta' \n",
    "#     f_cntr = f'P{ti}_C_peaks.fasta' \n",
    "#     result = subprocess.run([\"ame\", \"--oc\", ame_out,\n",
    "#                                  \"--control\", f_cntr, f_test, f_motifs,\n",
    "#                                ], cwd=wkdir, stderr=subprocess.PIPE, text=True)\n",
    "#     print(f\"Return code: {result.returncode}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fde53bc1-baed-4735-b850-74fd0ca2f42c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
